go-controller/pkg/ovn/ovn.go

```diff
@@ -1047,148 +1064,54 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
-	var mgmtPortFailed sync.Map
-	var addNodeFailed sync.Map
-	var nodeClusterRouterPortFailed sync.Map
+
+	go func() {
+		// track the retryNode map and every 30 seconds check if any node need to be retried
+		for {
+			select {
+			case <-time.After(30 * time.Second):
+				oc.iterateRetryNodes(false)
+			case <-oc.retryNodeChan:
+				oc.iterateRetryNodes(true)
+			case <-oc.stopChan:
+				return
+			}
+		}
+	}()
 
 	start := time.Now()
 	oc.watchFactory.AddNodeHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			node := obj.(*kapi.Node)
-			if noHostSubnet := noHostSubnet(node); noHostSubnet {
-				err := oc.lsManager.AddNoHostSubnetNode(node.Name)
-				if err != nil {
-					klog.Errorf("Error creating logical switch cache for node %s: %v", node.Name, err)
-				}
-				return
-			}
-
-			klog.V(5).Infof("Added event for Node %q", node.Name)
-			hostSubnets, err := oc.addNode(node)
-			if err != nil {
-				klog.Errorf("NodeAdd: error creating subnet for node %s: %v", node.Name, err)
-				addNodeFailed.Store(node.Name, true)
-				nodeClusterRouterPortFailed.Store(node.Name, true)
-				mgmtPortFailed.Store(node.Name, true)
-				gatewaysFailed.Store(node.Name, true)
+			oc.initRetryNode(node)
+			if err := oc.addNodeObj(node); err != nil {
+				klog.Errorf("Failed to create node %s, error: %v",
+					node.Name, err)
+				oc.unSkipRetryNode(node)
 				return
 			}
-
-			if err = oc.syncNodeClusterRouterPort(node, hostSubnets); err != nil {
-				if !util.IsAnnotationNotSetError(err) {
-					klog.Warningf(err.Error())
-				}
-				nodeClusterRouterPortFailed.Store(node.Name, true)
-			}
-
-			err = oc.syncNodeManagementPort(node, hostSubnets)
-			if err != nil {
-				if !util.IsAnnotationNotSetError(err) {
-					klog.Warningf("Error creating management port for node %s: %v", node.Name, err)
-				}
-				mgmtPortFailed.Store(node.Name, true)
-			}
-
-			if err := oc.syncNodeGateway(node, hostSubnets); err != nil {
-				if !util.IsAnnotationNotSetError(err) {
-					klog.Warningf(err.Error())
-				}
-				gatewaysFailed.Store(node.Name, true)
-			}
-
-			// ensure pods that already exist on this node have their logical ports created
-			options := metav1.ListOptions{FieldSelector: fields.OneTermEqualSelector("spec.nodeName", node.Name).String()}
-			pods, err := oc.client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), options)
-			if err != nil {
-				klog.Errorf("Unable to list existing pods on node: %s, existing pods on this node may not function")
-			} else {
-				oc.addRetryPods(pods.Items)
-				oc.requestRetryPods()
-			}
-
+			oc.checkAndDeleteRetryNode(node)
 		},
 		UpdateFunc: func(old, new interface{}) {
 			oldNode := old.(*kapi.Node)
 			node := new.(*kapi.Node)
-
-			shouldUpdate, err := shouldUpdate(node, oldNode)
-			if err != nil {
-				klog.Errorf(err.Error())
-			}
-			if !shouldUpdate {
-				// the hostsubnet is not assigned by ovn-kubernetes
+			if err := oc.updateNodeObj(node, oldNode); err != nil {
```

Comment:
before executing updateNodeObj we might need to fetch the newest version of node. In pod retry we do this, in policy we don't (might need to change that). The reason is in retry goroutine we fetch the latest version of the object. That means we could have revisions A,B,C (C is current). A fails to get added, goes into retry. Retry loop executes and adds C. Meanwhile B was added to the queue and then gets processed here. We dont want to update to B here. We want to get C and realize there is nothing to change.